### PR TITLE
Implement release channels in the launcher.

### PR DESCRIPTION
**Pull request by ThatRedox**
Replace the download snapshots checkmark box with a channel selection dropdown. Allow updating if the version is not already downloaded, changed from only if it is a chronologically newer version. Bump launcher version to 1.13.0. Add a launcher settings revision field for possible future use.
image
image

### CODE REVIEW

**Comment by ThatRedox** [comment]
Implement launcher version checking on startup.
image

**Comment by ThatRedox** [comment]
New features require the update site to have a `launcher.json` such as:
```json
{
  "name": "1.13.0",
  "timestamp": "2021-12-04T06:48:21Z",
  "notes": "Thanks to all the ...",
  "path": "ChunkyLauncher.jar",
  "channels": [
    {
      "name": "Stable",
      "path": "latest.json",
      "notes": "Latest stable version of Chunky."
    },
    {
      "name": "Beta",
      "path": "snapshot-stable.json",
      "notes": "Latest beta version of Chunky. Snapshots of next stable patch or release candidates."
    },
    {
      "name": "Snapshot",
      "path": "snapshot.json",
      "notes": "Latest nightly snapshot of Chunky."
    }
  ]
}
```
Old update sites are still compatible.

File: launcher/src/se/llbit/chunky/launcher/LauncherInfoChecker.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        } catch (MalformedURLException e) {
+            System.err.println("Malformed launcher info URL.");
+            errorListener.accept("Malformed launcher info/update site URL: " + url);
+        } catch (IOException e) {
+            System.err.println("Failed to fetch launcher info: " + e.getMessage());
+            errorListener.accept("Failed to fetch launcher info from URL: " + url);
+        } catch (SyntaxError e) {
+            System.err.println("Version info JSON error: " + e.getMessage());
+            errorListener.accept("Downloaded corrupt launcher info.");
+        } catch (Throwable e) {
```
Comment (leMaik) [inline launcher/src/se/llbit/chunky/launcher/LauncherInfoChecker.java:63]: If we get a `RuntimeException` or just any `Throwable` here, something is seriously wrong… we shouldn't handle that.

```suggestion
        } catch (Exception e) {
```

File: launcher/src/se/llbit/chunky/launcher/ui/ChunkyLauncherController.java [hunk 2]
```diff
@@ -188,14 +180,37 @@ public ChunkyLauncherController(LauncherSettings settings) {
...
+        });
+    updateChannelsList();
+
+    releaseChannelReload.setOnAction(event -> {
+      if (isBusy()) {
+        setBusy(true);
+        updateLauncher(
+            error -> Platform.runLater(() -> {
+              setBusy(false);
+              launcherError("Failed to Update", error);
```
Comment (leMaik) [inline launcher/src/se/llbit/chunky/launcher/ui/ChunkyLauncherController.java:204]:
```suggestion
              launcherError("Failed to update the launcher", error);
```

File: lib/src/se/llbit/chunky/launcher/LauncherSettings.java [hunk 3]
```diff
@@ -81,7 +100,37 @@ public void load() {
     version = settings.getString("version", "latest");
     showLauncher = settings.getBool("showLauncher", true);
     showAdvancedSettings = settings.getBool("showAdvancedSettings", false);
-    downloadSnapshots = settings.getBool("downloadSnapshots", false);
+
+    JsonObject releaseChannelObj = settings.get("releaseChannels").object();
+    releaseChannels = new ArrayList<>();
+    // Add these channels to support legacy update sites.
+    releaseChannels.add(STABLE_RELEASE_CHANNEL);
+    releaseChannels.add(SNAPSHOT_RELEASE_CHANNEL);
```
Comment (leMaik) [inline lib/src/se/llbit/chunky/launcher/LauncherSettings.java:108]: Would it be enough to only add these if we can't find a `launcher.json`?

File: lib/src/se/llbit/chunky/launcher/LauncherSettings.java [hunk 4]
```diff
@@ -105,7 +156,18 @@ public void save() {
     settings.setString("version", version);
     settings.setBool("showLauncher", showLauncher);
     settings.setBool("showAdvancedSettings", showAdvancedSettings);
-    settings.setBool("downloadSnapshots", downloadSnapshots);
+
+    settings.removeSetting("downloadSnapshots");
+    JsonObject releaseChannelsObject = new JsonObject();
+    int selectedChannelValue = releaseChannels.indexOf(selectedChannel);
+    if (selectedChannelValue == -1) selectedChannelValue = 0;
+    releaseChannelsObject.set("selectedChannel", new JsonNumber(selectedChannelValue));
```
Comment (leMaik) [inline lib/src/se/llbit/chunky/launcher/LauncherSettings.java:164]: Instead of storing the index, why not assign IDs to the channels?

**Review by leMaik — CHANGES_REQUESTED** [review]

**Pull request merged**